### PR TITLE
Added checkout subcommand

### PR DESCRIPTION
**Pull request by StefanBossbaly**
This command allows for the checking out of a topic branch across multiple repos. When working on a change that effects multiple repos, the general flow is to create a branch with the same name for ever repo that the change touches. These changes are then sent to gerrit and linked together under a topic. When landing, these changes are applied to their respective repos atomically. This change adds the checkout subcommand which allows the developer to change topic branches in their repo checkout easily.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tamird/pore/1)

### CODE REVIEW

**Comment by StefanBossbaly** [comment]
@tamird Looks like the workflow requires approval. Can I haz maintainer access to this repo?

i-can-has-gimme

**Comment by tamird** [comment]
Removed the approval requirement (I hope).

**Review by StefanBossbaly — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1#-:-O74nRHT-2886ztTJ4Vf:bnfp4nl)* status: 0 of 4 files reviewed, all discussions resolved

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: all files reviewed, 6 unresolved discussions (waiting on @StefanBossbaly)

___
*[`-- commits` line 17 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74YQTV4AZyWxCby0TU):*
Maybe better to just reference [https://gerrit.googlesource.com/git-repo/+/refs/heads/main/subcmds/branches.py](https://gerrit.googlesource.com/git-repo/+/refs/heads/main/subcmds/checkout.py)?
___
*[`src/main.rs` line 150 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74ZdkjAf7kOjxX0djy:-O74ZdkjAf7kOjxX0djz:bk736oi) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L150)):*
> ```XML
> 
>   /// Checkout a branch for development
>   Checkout {
> ```

did you want to keep this alphabetical? it should be above `Clone` (and you have it in that order in the implementation below)
___
*[`src/main.rs` line 151 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74YXs4581yotf_DFpD:-O74YXs4581yotf_DFpE:bviyoul) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L151)):*
> ```XML
>   /// Checkout a branch for development
>   Checkout {
>     /// Specify a branch to fetch (can be used multiple times)
> ```

this says fetch, but the above says checkout. which is it?
___
*[`src/main.rs` line 992 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74Zspm8nW5C1H-WFSK:-O74Zspm8nW5C1H-WFSL:b-1ivxpo) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L992)):*
> ```XML
>       Commands::Checkout { branch } => {
>         let tree = Tree::find_from_path(cwd)?;
>         cmd_checkout(config, &mut pool, &tree, &branch)
> ```

can we just inline this here? the one-line function isn't worth the indirection here. This seems to be following the lead of `cmd_clone` which is considerably more complex.
___
*[`src/tree.rs` line 1105 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74p7TX-KW7f1EFHz2p:-O74p7TX-KW7f1EFHz2q:bsvoqut) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1105)):*
> ```XML
>       eprintln!("{}: {}", error.name, error.result);
>       failed = true;
>     }
> ```

elsewhere there's a pattern like
```
  if !results.failed.is_empty() {
    for error in results.failed {
      eprintln!("{}: {}", error.name, error.result);
    }
    bail!("failed to git status");
  }
```
wouldn't that be simpler? you could replace `bail!` with `return 1`



_Code quote:_
```XML
    for error in results.failed {
      eprintln!("{}: {}", error.name, error.result);
      failed = true;
    }
```
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O74pcGT6b4o6ifMDU_P:b321ha4) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
> ```XML
> 
>     if !failed {
>       let checkout_projects = results
> ```

you should be able to do this without materializing a vector.

**Review by StefanBossbaly — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @tamird)

___
*[`-- commits` line 17 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74YQTV4AZyWxCby0TU):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Maybe better to just reference [https://gerrit.googlesource.com/git-repo/+/refs/heads/main/subcmds/branches.py](https://gerrit.googlesource.com/git-repo/+/refs/heads/main/subcmds/checkout.py)?
</blockquote></details>

Added link back to Google repo tool
___
*[`src/main.rs` line 150 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74ZdkjAf7kOjxX0djy:-O74wWjj19wrae0gOHIi:b-oqkyz) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did you want to keep this alphabetical? it should be above `Clone` (and you have it in that order in the implementation below)
</blockquote></details>

Reordered
___
*[`src/main.rs` line 151 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74YXs4581yotf_DFpD:-O74wxvV6p8Rwz0SNFsY:bt1q8lc) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L151)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this says fetch, but the above says checkout. which is it?
</blockquote></details>

Removed comment about fetching, that is handled in sync
___
*[`src/main.rs` line 992 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74Zspm8nW5C1H-WFSK:-O74xFyBCzepv-m4MZyD:bb6un1j) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L992)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can we just inline this here? the one-line function isn't worth the indirection here. This seems to be following the lead of `cmd_clone` which is considerably more complex.
</blockquote></details>

Done
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O74ypLx1t9Khb1RPf1B:b-2himcr) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you should be able to do this without materializing a vector.
</blockquote></details>

Fixed with using a peekable iterator

**Review by StefanBossbaly — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @tamird)

___
*[`src/tree.rs` line 1105 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74p7TX-KW7f1EFHz2p:-O75-72e5IxeEEcJDuHb:b-886gsx) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

elsewhere there's a pattern like
```
  if !results.failed.is_empty() {
    for error in results.failed {
      eprintln!("{}: {}", error.name, error.result);
    }
    bail!("failed to git status");
  }
```
wouldn't that be simpler? you could replace `bail!` with `return 1`
</blockquote></details>

Fixed

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @StefanBossbaly)

___
*[`src/main.rs` line 992 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74Zspm8nW5C1H-WFSK:-O751ZYC0Sf-pTUch-HJ:b-r0ywis) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L992)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

Done
</blockquote></details>

sorry, I meant inline the body of `cmd_checkeout`. Why did arc come back?
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O751f5M8SQmFSziVBf8:bwihm8l) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

Fixed with using a peekable iterator
</blockquote></details>

Why does it need to be peekable? just use the `let mut failed = false;` pattern and iterate over it just once.
___
*[`src/tree.rs` line 1115 at r7](https://reviewable.io/reviews/tamird/pore/1#-O751nvB0KBcvwNznnHY:-O751nvB0KBcvwNznnHZ:b1gqkbb) ([raw file](https://github.com/tamird/pore/blob/64013b24fc96853afcb328d3cfcb8b0347f4cece/src/tree.rs#L1115)):*
> ```XML
>     let results = pool.execute(job);
> 
>     if !results.failed.is_empty() {
> ```

this is backwards

**Review by StefanBossbaly — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @tamird)

___
*[`src/main.rs` line 992 at r4](https://reviewable.io/reviews/tamird/pore/1#-O74Zspm8nW5C1H-WFSK:-O754fOtALigzj0N-CcU:b-cmp0jn) ([raw file](https://github.com/tamird/pore/blob/f6a49d96f2a2c69ff6894896d2c603855c63bba8/src/main.rs#L992)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

sorry, I meant inline the body of `cmd_checkeout`. Why did arc come back?
</blockquote></details>

Because I switch to my mac. Rebasing
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O754oj08PpGvLTV97rV:blg7py9) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Why does it need to be peekable? just use the `let mut failed = false;` pattern and iterate over it just once.
</blockquote></details>

I feel that using `peekable` is better than having a lambda with side effects.
___
*[`src/tree.rs` line 1115 at r7](https://reviewable.io/reviews/tamird/pore/1#-O751nvB0KBcvwNznnHY:-O754lHA3C8GaEKD6uYW:b-lg9zyx) ([raw file](https://github.com/tamird/pore/blob/64013b24fc96853afcb328d3cfcb8b0347f4cece/src/tree.rs#L1115)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is backwards
</blockquote></details>

More words please

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @StefanBossbaly)

___
*[`src/main.rs` line 457 at r9](https://reviewable.io/reviews/tamird/pore/1#-O75dgel3OYlsN2Hz8Y-:-O75dgel3OYlsN2Hz8Y0:bo4iy2i) ([raw file](https://github.com/tamird/pore/blob/525d672ea03c6b93e646da3a8af38ea927e756c5/src/main.rs#L457)):*
> ```XML
> }
> 
> fn cmd_checkout(config: &Config, pool: &mut Pool, tree: &Tree, target_branch: &str) -> Result<i32, Error> {
> ```

inline this please - one line methods not useful
___
*[`src/main.rs` line 990 at r9](https://reviewable.io/reviews/tamird/pore/1#-O75dkkvB-cXY2MCA6IS:-O75dkkvB-cXY2MCA6IT:b8zrobm) ([raw file](https://github.com/tamird/pore/blob/525d672ea03c6b93e646da3a8af38ea927e756c5/src/main.rs#L990)):*
> ```XML
>         )
>       }
>       Commands::Checkout { branch } => cmd_checkout(&config, &mut pool, &Tree::find_from_path(cwd)?, &branch),
> ```

separate expression for this was better



_Code quote:_
```XML
Tree::find_from_path(cwd)?
```
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O75drew27BLnqDgcJxa:b-ijusji) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

I feel that using `peekable` is better than having a lambda with side effects.
</blockquote></details>

Um. What?

```
      let mut found = false;
      for checkout_project in results.successful.filter_map(|result| result.result) {
        found = true;
        println!("Checked out {checkout_project}");
      }
      if found {
        Ok(0)
      } else {
        eprintln!("error: no project has branch {target_branch}");
        Ok(1)
      }
```

___
*[`src/tree.rs` line 1115 at r7](https://reviewable.io/reviews/tamird/pore/1#-O751nvB0KBcvwNznnHY:-O75dnJl3V2mq1qytfD7:b-bjodt5) ([raw file](https://github.com/tamird/pore/blob/64013b24fc96853afcb328d3cfcb8b0347f4cece/src/tree.rs#L1115)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

More words please
</blockquote></details>

my mistake, is correct

**Review by StefanBossbaly — COMMENTED** [review]
Addressed comments

> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1#-:-O75lTD9-DCscoE0nd1b:b-d9w856)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @tamird)

___
*[`src/main.rs` line 457 at r9](https://reviewable.io/reviews/tamird/pore/1#-O75dgel3OYlsN2Hz8Y-:-O75izQaByATQVXTop1U:bncjm0f) ([raw file](https://github.com/tamird/pore/blob/525d672ea03c6b93e646da3a8af38ea927e756c5/src/main.rs#L457)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

inline this please - one line methods not useful
</blockquote></details>

I agree but was trying to keep the form of other functions, since they are all dispatched from `cmd_*` style functions. Will change but I think there is a benefit to keep it as-is since it allows developer to easily switch between cmd functions
___
*[`src/main.rs` line 990 at r9](https://reviewable.io/reviews/tamird/pore/1#-O75dkkvB-cXY2MCA6IS:-O75iqEX4vHtDENvN10Z:bxio2sp) ([raw file](https://github.com/tamird/pore/blob/525d672ea03c6b93e646da3a8af38ea927e756c5/src/main.rs#L990)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

separate expression for this was better
</blockquote></details>

Confusing this with the previous `cmd_checkout` suggestion, reverted
___
*[`src/tree.rs` line 1108 at r5](https://reviewable.io/reviews/tamird/pore/1#-O74pcGT6b4o6ifMDU_O:-O75jJRz-cjMZUKZ6C5o:bfbk1wd) ([raw file](https://github.com/tamird/pore/blob/050550e05906a9e2acb0bd8e2cb60255111d5f5f/src/tree.rs#L1108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Um. What?

```
      let mut found = false;
      for checkout_project in results.successful.filter_map(|result| result.result) {
        found = true;
        println!("Checked out {checkout_project}");
      }
      if found {
        Ok(0)
      } else {
        eprintln!("error: no project has branch {target_branch}");
        Ok(1)
      }
```

</blockquote></details>

I thought you wanted the bool setting in the filter_map lambda. Changing...

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @StefanBossbaly)

___
*[`src/main.rs` line 457 at r9](https://reviewable.io/reviews/tamird/pore/1#-O75dgel3OYlsN2Hz8Y-:-O75mdBJBdnlfwcqEOdP:b-if0fba) ([raw file](https://github.com/tamird/pore/blob/525d672ea03c6b93e646da3a8af38ea927e756c5/src/main.rs#L457)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

I agree but was trying to keep the form of other functions, since they are all dispatched from `cmd_*` style functions. Will change but I think there is a benefit to keep it as-is since it allows developer to easily switch between cmd functions
</blockquote></details>

As far as I can tell (and I said this in an earlier comment) there is only one `cmd_` function and it is way more complex than the other commands.
___
*[`src/tree.rs` line 1064 at r10](https://reviewable.io/reviews/tamird/pore/1#-O75mvrG9I8V_EVhQ-zt:-O75mvrG9I8V_EVhQ-zu:b-d7gpos) ([raw file](https://github.com/tamird/pore/blob/01018e2a58dc43f0d7febeffcec09cf5cfe6ddb9/src/tree.rs#L1064)):*
> ```XML
>     for project in projects {
>       job.add_task(
>         project.project_path.clone(),
> ```

you can avoid cloning here if you iterate over refs instead (and you don't need a move closure since you don't return anything borrowed from `projects` from this function)

**Review by StefanBossbaly — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @tamird)

___
*[`src/tree.rs` line 1064 at r10](https://reviewable.io/reviews/tamird/pore/1#-O75mvrG9I8V_EVhQ-zt:-O75p9fx02Dd-ijp259z:be0jh5q) ([raw file](https://github.com/tamird/pore/blob/01018e2a58dc43f0d7febeffcec09cf5cfe6ddb9/src/tree.rs#L1064)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you can avoid cloning here if you iterate over refs instead (and you don't need a move closure since you don't return anything borrowed from `projects` from this function)
</blockquote></details>

Not sure if this is correct. The following will work:
```rust
let project = projects.get(0).unwrap();
job.add_task(&project.project_path, || -> Result<Option<&str>, Error> {
    let repo = git2::Repository::open(self.path.join(&project.project_path))
        .context(format!("failed to open object repository {:?}", project.project_path))?;
  ...
});

let results = pool.execute(job);
```

The problem that I am running into is that in the for loop the reference will not live long enough since 'task will outlive the reference which causes a borrow checker error:

```rust
for project in &projects {
  job.add_task(&project.project_path, || -> Result<Option<&str>, Error> {
    let repo = git2::Repository::open(self.path.join(&project.project_path))
      .context(format!("failed to open object repository {:?}", project.project_path))?;
  });
  // 'task will now outlive project, which causes an error
}
```

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 1 of 4 files reviewed, 1 unresolved discussion (waiting on @StefanBossbaly)

___
*[`src/tree.rs` line 1064 at r10](https://reviewable.io/reviews/tamird/pore/1#-O75mvrG9I8V_EVhQ-zt:-O75voPOBf7oi1CBNiSP:b36z6cg) ([raw file](https://github.com/tamird/pore/blob/01018e2a58dc43f0d7febeffcec09cf5cfe6ddb9/src/tree.rs#L1064)):*
<details><summary><i>Previously, StefanBossbaly (Stefan Bossbaly) wrote…</i></summary><blockquote>

Not sure if this is correct. The following will work:
```rust
let project = projects.get(0).unwrap();
job.add_task(&project.project_path, || -> Result<Option<&str>, Error> {
    let repo = git2::Repository::open(self.path.join(&project.project_path))
        .context(format!("failed to open object repository {:?}", project.project_path))?;
  ...
});

let results = pool.execute(job);
```

The problem that I am running into is that in the for loop the reference will not live long enough since 'task will outlive the reference which causes a borrow checker error:

```rust
for project in &projects {
  job.add_task(&project.project_path, || -> Result<Option<&str>, Error> {
    let repo = git2::Repository::open(self.path.join(&project.project_path))
      .context(format!("failed to open object repository {:?}", project.project_path))?;
  });
  // 'task will now outlive project, which causes an error
}
```
</blockquote></details>

```diff
diff --git a/src/tree.rs b/src/tree.rs
index c721b1d..5fa7062 100644
--- a/src/tree.rs
+++ b/src/tree.rs
@@ -1071,11 +1071,10 @@ impl Tree {
 
     let mut job = Job::with_name("checkout");
 
-    for project in projects {
-      job.add_task(
-        project.project_path.clone(),
-        move || -> Result<Option<String>, Error> {
-          let repo = git2::Repository::open(self.path.join(&project.project_path))
+    for project in &projects {
+      let path = self.path.join(&project.project_path);
+      job.add_task(&project.project_path, move || -> Result<Option<&str>, Error> {
+        let repo = git2::Repository::open(path)
           .context(format!("failed to open object repository {:?}", project.project_path))?;
 
         let maybe_parse = match repo.revparse_ext(target_branch) {
@@ -1097,12 +1096,11 @@ impl Tree {
             None => repo.set_head_detached(object.id())?,
           }
 
-            Ok(Some(project.project_path))
+          Ok(Some(&project.project_path))
         } else {
           Ok(None)
         }
-        },
-      );
+      });
     }
 
     let results = pool.execute(job);

```

**Review by StefanBossbaly — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1)* status: 1 of 4 files reviewed, 1 unresolved discussion (waiting on @tamird)

___
*[`src/tree.rs` line 1064 at r10](https://reviewable.io/reviews/tamird/pore/1#-O75mvrG9I8V_EVhQ-zt:-O75zyfa5Zs_uLCgsH3Y:b-fde5ak) ([raw file](https://github.com/tamird/pore/blob/01018e2a58dc43f0d7febeffcec09cf5cfe6ddb9/src/tree.rs#L1064)):*
>you can avoid cloning here if you iterate over refs instead (**and you don't need a move closure** since you don't return anything borrowed from `projects` from this function)

Ah I was trying to solve it without the move capture lamdba

**Review by tamird — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tamird/pore/1#-:-O761Thu4k_aBe35mAVI:bnfp4nl)* status: 2 of 4 files reviewed, all discussions resolved (waiting on @StefanBossbaly)

**Pull request merged**
